File: src/stores/checklists.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import type {
  Checklist,
  ChecklistSummary,
  CheckItem,
  CheckItemAbnormalDetail,
  PreCheckStatItem,
  MemberCompletionRankItem,
  FrequentAbnormalTypeItem,
} from '@/types'
import { useApi } from '@/composables/useApi'

export const useChecklistsStore = defineStore('checklists', () => {
  const api = useApi()
  const currentChecklist = ref<Checklist | null>(null)
  const summaries = ref<ChecklistSummary[]>([])
  const abnormalItems = ref<CheckItemAbnormalDetail[]>([])
  const preCheckStats = ref<PreCheckStatItem[]>([])
  const memberRank = ref<MemberCompletionRankItem[]>([])
  const abnormalTypes = ref<FrequentAbnormalTypeItem[]>([])
  const loading = ref(false)

  async function generateChecklist(performanceId: number, items?: { song_id: number; category: string; item_name: string; responsible_member_id?: number; position_id?: string; deadline?: string }[]) {
    loading.value = true
    try {
      currentChecklist.value = await api.checklists.generate(performanceId, items)
    } catch (e) {
      console.error('Failed to generate checklist', e)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function fetchChecklist(performanceId: number) {
    loading.value = true
    try {
      currentChecklist.value = await api.checklists.getByPerformance(performanceId)
    } catch (e) {
      console.error('Failed to fetch checklist', e)
      currentChecklist.value = null
    } finally {
      loading.value = false
    }
  }

  async function fetchSummary(performanceId: number) {
    try {
      return await api.checklists.getSummary(performanceId)
    } catch (e) {
      console.error('Failed to fetch summary', e)
      return null
    }
  }

  async function fetchAbnormalItems(performanceId: number) {
    try {
      abnormalItems.value = await api.checklists.getAbnormalItems(performanceId)
    } catch (e) {
      console.error('Failed to fetch abnormal items', e)
      abnormalItems.value = []
    }
  }

  async function fetchMemberItems(performanceId: number, memberId: number) {
    try {
      return await api.checklists.getMemberItems(performanceId, memberId)
    } catch (e) {
      console.error('Failed to fetch member items', e)
      return []
    }
  }

  async function updateItem(itemId: number, data: { status?: string; abnormal_description?: string; photo_url?: string }) {
    const result = await api.checklists.updateItem(itemId, data)
    if (currentChecklist.value) {
      const idx = currentChecklist.value.items.findIndex((i) => i.id === itemId)
      if (idx >= 0) {
        currentChecklist.value.items[idx] = result
      }
    }
    return result
  }

  async function addItem(performanceId: number, data: { song_id: number; category: string; item_name: string; responsible_member_id?: number; position_id?: string; deadline?: string }) {
    const result = await api.checklists.addItem(performanceId, data)
    if (currentChecklist.value) {
      currentChecklist.value.items.push(result)
    }
    return result
  }

  async function deleteItem(itemId: number) {
    await api.checklists.deleteItem(itemId)
    if (currentChecklist.value) {
      currentChecklist.value.items = currentChecklist.value.items.filter((i) => i.id !== itemId)
    }
  }

  async function fetchAllSummaries() {
    try {
      summaries.value = await api.checklists.getAllSummaries()
    } catch (e) {
      console.error('Failed to fetch summaries', e)
    }
  }

  async function fetchPreCheckStats() {
    try {
      preCheckStats.value = await api.checklists.getPreCheckStats()
    } catch (e) {
      console.error('Failed to fetch pre-check stats', e)
    }
  }

  async function fetchMemberRank() {
    try {
      memberRank.value = await api.checklists.getMemberRank()
    } catch (e) {
      console.error('Failed to fetch member rank', e)
    }
  }

  async function fetchAbnormalTypes() {
    try {
      abnormalTypes.value = await api.checklists.getAbnormalTypes()
    } catch (e) {
      console.error('Failed to fetch abnormal types', e)
    }
  }

  async function fetchAllStats() {
    loading.value = true
    try {
      await Promise.all([
        fetchAllSummaries(),
        fetchPreCheckStats(),
        fetchMemberRank(),
        fetchAbnormalTypes(),
      ])
    } finally {
      loading.value = false
    }
  }

  return {
    currentChecklist,
    summaries,
    abnormalItems,
    preCheckStats,
    memberRank,
    abnormalTypes,
    loading,
    generateChecklist,
    fetchChecklist,
    fetchSummary,
    fetchAbnormalItems,
    fetchMemberItems,
    updateItem,
    addItem,
    deleteItem,
    fetchAllSummaries,
    fetchPreCheckStats,
    fetchMemberRank,
    fetchAbnormalTypes,
    fetchAllStats,
  }
})

File: api/routers/checklists.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from database import get_db
from schemas import (
    ChecklistGenerateRequest,
    ChecklistResponse,
    ChecklistSummaryResponse,
    CheckItemCreate,
    CheckItemUpdate,
    CheckItemResponse,
    CheckItemAbnormalDetail,
    PreCheckStatItem,
    MemberCompletionRankItem,
    FrequentAbnormalTypeItem,
)
import services

router = APIRouter(prefix="/api/checklists", tags=["checklists"])


@router.post("/performances/{performance_id}", response_model=ChecklistResponse)
def generate_checklist(performance_id: int, data: ChecklistGenerateRequest, db: Session = Depends(get_db)):
    result = services.generate_checklist(db, performance_id, data.model_dump())
    if not result:
        raise HTTPException(status_code=404, detail="Performance task not found or checklist already exists")
    return result


@router.get("/performances/{performance_id}", response_model=ChecklistResponse)
def get_checklist(performance_id: int, db: Session = Depends(get_db)):
    result = services.get_checklist_by_performance(db, performance_id)
    if not result:
        raise HTTPException(status_code=404, detail="Checklist not found")
    return result


@router.get("/performances/{performance_id}/summary", response_model=ChecklistSummaryResponse)
def get_summary(performance_id: int, db: Session = Depends(get_db)):
    result = services.get_checklist_summary(db, performance_id)
    if not result:
        raise HTTPException(status_code=404, detail="Checklist not found")
    return result


@router.get("/performances/{performance_id}/abnormal", response_model=list[CheckItemAbnormalDetail])
def get_abnormal(performance_id: int, db: Session = Depends(get_db)):
    return services.get_abnormal_items(db, performance_id)


@router.get("/performances/{performance_id}/member/{member_id}", response_model=list[CheckItemResponse])
def get_member_items(performance_id: int, member_id: int, db: Session = Depends(get_db)):
    return services.get_member_check_items(db, performance_id, member_id)


@router.put("/items/{item_id}", response_model=CheckItemResponse)
def update_item(item_id: int, data: CheckItemUpdate, db: Session = Depends(get_db)):
    result = services.update_check_item(db, item_id, data.model_dump(exclude_unset=True))
    if not result:
        raise HTTPException(status_code=404, detail="Check item not found")
    return result


@router.post("/performances/{performance_id}/items", response_model=CheckItemResponse)
def add_item(performance_id: int, data: CheckItemCreate, db: Session = Depends(get_db)):
    result = services.add_check_item(db, performance_id, data.model_dump())
    if not result:
        raise HTTPException(status_code=404, detail="Checklist not found")
    return result


@router.delete("/items/{item_id}")
def delete_item(item_id: int, db: Session = Depends(get_db)):
    success = services.delete_check_item(db, item_id)
    if not success:
        raise HTTPException(status_code=404, detail="Check item not found")
    return {"success": True}


@router.get("/summaries", response_model=list[ChecklistSummaryResponse])
def get_all_summaries(db: Session = Depends(get_db)):
    return services.get_all_checklist_summaries(db)


@router.get("/stats/pre-check", response_model=list[PreCheckStatItem])
def get_pre_check_stats(db: Session = Depends(get_db)):
    return services.get_pre_check_stats(db)


@router.get("/stats/member-rank", response_model=list[MemberCompletionRankItem])
def get_member_rank(db: Session = Depends(get_db)):
    return services.get_member_completion_rank(db)


@router.get("/stats/abnormal-types", response_model=list[FrequentAbnormalTypeItem])
def get_abnormal_types(db: Session = Depends(get_db)):
    return services.get_frequent_abnormal_types(db)
